core/grid.js

```diff
@@ -0,0 +1,221 @@
+/**
+ * @license
+ * Visual Blocks Editor
+ *
+ * Copyright 2014 Google Inc.
+ * https://developers.google.com/blockly/
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview Object for configuring and updating a workspace grid in
+ * Blockly.
+ * @author fenichel@google.com (Rachel Fenichel)
+ */
+'use strict';
+
+goog.provide('Blockly.Grid');
+
+goog.require('Blockly.utils');
+
+goog.require('goog.userAgent');
+
+
+/**
+ * Class for a workspace's grid.
+ * @param {!SVGElement} pattern The grid's SVG pattern, created during injection.
+ * @param {!Object} options A dictionary of normalized options for the grid.
+ *     See grid documentation:
+ *     https://developers.google.com/blockly/guides/configure/web/grid
+ * @constructor
+ */
+Blockly.Grid = function(pattern, options) {
+  /**
+   * The grid's SVG pattern, created during injection.
+   * @type {!SVGElement}
+   * @private
+   */
+  this.gridPattern_ = pattern;
+
+  /**
+   * The spacing of the grid lines.
+   * @type {number}
+   * @private
+   */
+  this.spacing_ = options['spacing'];
+
+  /**
+   * How long the grid lines should be.
+   * @type {number}
+   * @private
+   */
+  this.length_ = options['length'];
+
+  /**
+   * The horizontal grid line, if it exists.
+   * @type {SVGElement}
+   * @private
+   */
+  this.line1_ = pattern.firstChild;
+
+  /**
+   * The vertical grid line, if it exists.
+   * @type {SVGElement}
+   * @private
+   */
+  this.line2_ = this.line1_ && this.line1_.nextSibling;
+
+  /**
+   * Whether blocks should snap to the grid.
+   * @type {boolean}
+   * @private
+   */
+  this.snapToGrid_ = options['snap'];
+};
+
+/**
+ * The scale of the grid, used to set stroke width on grid lines.
+ * @type {number}
+ * @private
+ */
+Blockly.Grid.prototype.scale_ = 1;
+
+/**
+ * Dispose of this grid and unlink from the DOM.
+ * @package
+ */
+Blockly.Grid.prototype.dispose = function() {
+  this.gridPattern_ = null;
+};
+
+/**
+ * Whether blocks should snap to the grid, based on the initial configuration.
+ * @return {boolean} True if blocks should snap, false otherwise.
```

Comment:
No, because according to the closure type system rules, "Note that the following are already non-nullable, and thus prepending! is redundant: !number, !string, !boolean
!{foo: string}, !function()"

It's allowed but not required.